PC/pyconfig.h

```diff
@@ -72,12 +72,18 @@ WIN32 is still required for the locale module.
 #define USE_SOCKET
 #endif
 
-#if defined(WINAPI_FAMILY) && (WINAPI_FAMILY != WINAPI_FAMILY_DESKTOP_APP)
-#define MS_WINDOWS_NON_DESKTOP
-#endif
-
-#if defined(WINAPI_FAMILY) && (WINAPI_FAMILY == WINAPI_FAMILY_GAMES)
-#define MS_WINDOWS_GAMES
+#if !defined(WINAPI_FAMILY) || (WINAPI_FAMILY == WINAPI_FAMILY_DESKTOP_APP)
```

Comment:
The SDK uses API partitions instead of API families:

```text
WINAPI_PARTITION_DESKTOP            // usable for Desktop Win32 apps (but not store apps)
WINAPI_PARTITION_APP                // usable for Windows Universal store apps
WINAPI_PARTITION_PC_APP             // specific to Desktop-only store apps
WINAPI_PARTITION_PHONE_APP          // specific to Phone-only store apps
WINAPI_PARTITION_SYSTEM             // specific to System applications 
WINAPI_PARTITION_SERVER
WINAPI_PARTITION_GAMES  
```

For example, `WINAPI_PARTITION_APP` includes the desktop app, PC app, and phone app families. The desktop app family is member of all partitions except for system and server.

Reply:
Yes but it is unclear to me how to set them in a "meaningful" way for us here. E.g. `WINAPI_PARTITION_GAMES` is `WINAPI_FAMILY_DESKTOP` + `WINAPI_FAMILY_GAMES`.

Reply:
I know the table. However they do not follow their own spec ...
```
#ifndef WINAPI_PARTITION_GAMES
#define WINAPI_PARTITION_GAMES                                                 \
  (WINAPI_FAMILY == WINAPI_FAMILY_GAMES ||                                     \
   WINAPI_FAMILY == WINAPI_FAMILY_DESKTOP_APP)
#endif
```

Reply:
I'm not at all surprised it isn't followed precisely...

Part of the problem is that there isn't really "Xbox related code" from our POV. There's "code that is only required when the family is [not] GAMES" and "code that should be omitted when the family is [not] GAMES".

I think rephrasing those in terms of partitions is less readable.

Reply:
Help me out here. I don't see a fundamental problem with doing what the SDK does, and I'd prefer to match what they do since it makes it simpler to verify correctness when comparing the SDK and Python's source code.

```c
#if WINAPI_FAMILY_PARTITION(WINAPI_PARTITION_DESKTOP)
#define MS_WINDOWS_DESKTOP
#endif
#if WINAPI_FAMILY_PARTITION(WINAPI_PARTITION_APP)
#define MS_WINDOWS_APP
#endif
#if WINAPI_FAMILY_PARTITION(WINAPI_PARTITION_SYSTEM)
#define MS_WINDOWS_SYSTEM
#endif
#if WINAPI_FAMILY_PARTITION(WINAPI_PARTITION_GAMES)
#define MS_WINDOWS_GAMES
#endif
#if defined(MS_WINDOWS_APP) || defined(MS_WINDOWS_SYSTEM)
#define HAVE_WINDOWS_CONSOLE_IO 1
#endif
```

If we need `SEM_NOGPFAULTERRORBOX` as in "Modules/faulthandler.c", we'd gate the code with `defined(MS_WINDOWS_DESKTOP)`.

If we need the ANSI functions `OpenFileMappingA()` and `CreateFileMappingA()`, as in "Modules/mmapmodule.c", we gate with `defined(MS_WINDOWS_DESKTOP) || defined(MS_WINDOWS_GAMES)`.

If we need the version helpers, as in "Modules/socketmodule.c", then we gate with `defined(MS_WINDOWS_DESKTOP)`.

If we need the registry API, it's mostly available if `defined(MS_WINDOWS_DESKTOP) || defined(MS_WINDOWS_SYSTEM) || defined(MS_WINDOWS_GAMES)`. Some registry functions such as `RegFlushKey()` are only declared if `defined(MS_WINDOWS_DESKTOP) || defined(MS_WINDOWS_SYSTEM)`.

If we need `GetConsoleCP()` and `GetConsoleOutputCP()`, as in "Python/fileutils.c", use a small amount of nesting, which for me doesn't hurt readability. 

```c
#ifdef MS_WINDOWS
#ifdef HAVE_WINDOWS_CONSOLE_IO
    // Windows console I/O code
#else
    Py_RETURN_NONE;
#endif
#else
    // POSIX code
#endif
```

Or similarly for the `Path*` functions:

```c
#ifdef MS_WINDOWS
#if defined(MS_WINDOWS_APP) || defined(MS_WINDOWS_SYSTEM)
    // Path API code
#else
    // fallback code
#endif
#else
    // POSIX code
#endif 
```

Reply:
The only issue is that we often need a workaround only on the playstation and
```
#if MS_WINDOWS_GAMES
  // "xbox" code
#else
  // normal code
```
does not work. Still after thinking about this a bit more it is probably still the better option, since it is the way the APIs are exposed by Microsoft.

Reply:
I think it's somewhat helped by having `MS_WINDOWS_DESKTOP` as opposed to the non-desktop, but my main dislike is that we could have multiple names defined at once. When using families, only one will be set.

But maybe it would work okay to just use the `WINAPI_FAMILY_PARTITION` macro directly? That at least has the advantage that we don't have to define/document it, and we can then use whichever logic makes the most sense in any circumstance. It would always have to be inside an `MS_WINDOWS` block, but it probably would be anyway.

Where I'm really stuck here is on the "how to teach this" question. The easier it is for someone to figure out how to use this API correctly, the less likely we'll introduce bugs in the future. So I want it to be easy to figure out - sometimes, that means making it appear so complicated that nobody tries to figure it out themselves, and using the full macro each time might achieve that.

Reply:
I just started writing the code for the Partition and I think it is clearer what it achieves. This is pretty much the same as directly using `WINAPI_FAMILY_PARTITION`. One problem I see with using `WINAPI_FAMILY_PARTITION` directly everywhere, is that this requires and additional check if we are on windows -> whether it is even defined WINAPI_FAMILY_PARTITION

Reply:
Updated the implementation to use partitions